### ISSUE TITLE
Proposal: Use lua-typecheck-action for all PRs

### ISSUE THREAD

**Issue by pysan3** [nvim-neorg/neorg#1160]
### Issues

- [X] I have checked [existing issues](https://github.com/vhyrro/neorg/issues?q=is%3Aissue) and there are no existing ones with the same request.

### Feature description

Recently I've been hearing about neorg more and more and I can see the project growing rapidly and more people contributing to the project.

Whilst that is definitely a good thing, it is inevitable that the codebase will become more messy and will potentially lead to more bugs.

https://github.com/mrcjkb/lua-typecheck-action is a GH Action which checks for type issues using `lua_ls` and I believe will be able to alert us on many potential issues that the holy untyped-stripting language might bring to us.
I suggest adding this action to this repo as well as adding more type notation into the codebase.

Acceptable type notations are listed here: https://github.com/LuaLS/lua-language-server/wiki/Annotations.

### Help

No

### Implementation help

Sorry I'm quite busy right now.

**Comment by vhyrro** [comment]
Yep this sounds like a fantastic idea! Should have some time either later today or tommorow :crossed_fingers:

**Comment by pysan3** [comment]
Sweet! Amazing work @vhyrro !

**Comment by pysan3** [comment]
Reference;

https://github.com/mrcjkb/lua-typecheck-action/issues/34

**Comment by max397574** [comment]
just wondering
do we want to use the neodev types?
shouldn't the long term goal be to use the core ones?

**Comment by pysan3** [comment]
`neodev.nvim/types/{stable,nightly}` folder includes the type annotations for all vim core apis, and that's what we are referring to.

https://github.com/folke/neodev.nvim/blob/main/types/stable/api.lua

**Comment by max397574** [comment]
> neodev.nvim/types/{stable,nightly} folder includes the auto-generated type annotations for all vim core apis, and that's what I was referring to.
Not trying to use neodev's own functionality (neodev.nvim/lua) (in fact, I don't call require("neodev").setup in the action so neodev is doing mostly nothing.

I know
but most of those annotations are now also in core

**Comment by pysan3** [comment]
Oh, didn't know that!

That's a nice thing to have but looking at some issues, it's not completed all the way through? 
Also, that action is using the nix version of neovim ig? and I have no knowledge of nix but where can I find the `$XDG_DATA_HOME/nvim/runtime/lua` directory?

If we add that path to lua_ls, I think we can use that instead. I just don't know anything about nix and installing via neodev was easier.
At least I got undeclared global variable `vim` error when I used it without any parameter so I don't think it works ootb.
